lib/sql.ts

```diff
@@ -0,0 +1,208 @@
+		}
+		return this.query({type: 'get', data, num});
+	}
+	run(statement: string | number, data: DataType) {
+		const num = typeof statement === 'number' ? statement : this.statements.get(statement);
+		if (num === undefined || ![...this.statements.values()].includes(num)) {
+			throw new Error(`Prepare a statement before using another database function with SQLDatabase.prepare.`);
+		}
+		return this.query({type: 'run', num, data});
+	}
+	/** We don't want these to be prepared statements, since this should be used SPARINGLY. */
+	exec(statement: string) {
+		return this.query({type: 'exec', data: statement});
+	}
+	query(args: DatabaseQuery) {
+		this.process.send(args);
+		return new Promise<any>(resolve => {
+			this.pendingRequests.push(resolve);
+		});
+	}
+}
+
+class SQLProcessManager extends ProcessManager {
+	constructor(module: NodeJS.Module) {
+		super(module);
+	}
+	createProcess(options: SQLOptions) {
+		const process: ProcessWrapper = new DatabaseWrapper(options);
+		this.processes.push(process);
+		return process;
+	}
+	listen() {}
+	destroyProcess(process: DatabaseWrapper) {
+		void process.release();
+		this.processes.splice(this.processes.indexOf(process), 1);
+	}
+}
+
+export const PM = new SQLProcessManager(module);
+
+if (!PM.isParentProcess) {
+	let statementNum = 0;
+	const statements: Map<number, Sqlite.Statement> = new Map();
+	const transactions: Map<number, Sqlite.Transaction> = new Map();
+	const {file, extension} = process.env;
+	const database = new Sqlite(file!);
+	if (extension) {
+		// eslint-disable-next-line @typescript-eslint/no-var-requires
+		const {functions, storedTransactions} = require(`../${extension}`);
+		for (const k in functions) {
+			database.function(k, functions[k]);
+		}
+		for (const t in storedTransactions) {
+			const transaction = database.transaction(storedTransactions[t]);
+			transactions.set(transactions.size + 1, transaction);
+		}
+	}
+	database.pragma(`foreign_keys=on`);
+	process.on('message', (query: DatabaseQuery) => {
+		let statement;
+		let results;
+		switch (query.type) {
+		case 'prepare': {
+			const {data} = query;
+			const newStatement = database.prepare(data);
+			const nextNum = statementNum++;
+			statements.set(nextNum, newStatement);
+			return process.send!(nextNum);
+		}
+		case 'all': {
+			const {num, data} = query;
+			statement = statements.get(num);
+			results = statement?.all(data);
+		}
+			break;
+		case 'get': {
+			const {num, data} = query;
+			statement = statements.get(num);
+			results = statement?.get(data);
+		}
+			break;
+		case 'run': {
+			const {num, data} = query;
+			statement = statements.get(num);
+			results = statement?.run(data);
+		}
+			break;
+		case 'exec': {
+			const {data} = query;
+			database.exec(data);
+		}
+			break;
+		case 'transaction': {
+			const {num, data} = query;
+			const transaction = transactions.get(num);
+			if (!transaction) break;
+			results = transaction(data);
```

Comment:
I don't understand how this works -- presumably the transaction function needs to be passed some sort of connection object it can use? Are you envisioning the transaction function using the methods on `database`, and if so, how would you determine which transaction these methods are associated with?

Needs tests too

Reply:
I overlooked that. I'll probably pass it the connection parameter, yes.

Reply:
Done.